src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,760 @@
+    header then calls the appropriate smb_XXX
+    method with data beyond the header.
+
+    @param packet: the raw packet
+    @type packet: L{_base.SMBPacket}
+    """
+    offset = 0
+    isRelated = True
+    while isRelated:
+        protocol_id = packet.data[offset : offset + len(SMB2_MAGIC)]
+        if protocol_id == SMB1_MAGIC:
+            # its a SMB1 packet which we dont support with the exception
+            # of the first packet, we try to offer upgrade to SMB2
+            if packet.ctx.get("avatar") is None:
+                log.debug("responding to SMB1 packet")
+                negotiateResponse(packet)
+            else:
+                packet.close()
+                log.error("Got SMB1 packet while logged in")
+            return
+        elif protocol_id != SMB2_MAGIC:
+            packet.close()
+            log.error("Unknown packet type")
+            log.debug("packet data {data!r}", data=packet.data[offset : offset + 64])
+            return
+        packet.hdr, o2 = _base.unpack(HeaderSync, packet.data, offset, _base.OFFSET)
+        isAsync = (packet.hdr.flags & FLAG_ASYNC) > 0
+        isRelated = (packet.hdr.flags & FLAG_RELATED) > 0
+        isSigned = (packet.hdr.flags & FLAG_SIGNED) > 0
+        # FIXME other flags 3.1 or too obscure
+        if isAsync:
+            packet.hdr = _base.unpack(HeaderAsync, packet.data, offset)
+        if isRelated:
+            this_packet = packet.data[offset : offset + packet.hdr.next_command]
+        else:
+            this_packet = packet.data[offset:]
+        flags_desc = ""
+        if isAsync:
+            flags_desc += " ASYNC"
+        if isRelated:
+            flags_desc += " RELATED"
+        if isSigned:
+            flags_desc += " SIGNED"
+        log.debug(
+            """
+HEADER
+------
+protocol ID     {pid!r}
+size            {hs}
+credit charge   {cc}
+status          {status}
+command         {cmd!r} {cmdn:02x}
+credit request  {cr}
+flags           0x{flags:04x}{flags_desc}
+next command    0x{nc:x}
+message ID      0x{mid:x}
+session ID      0x{sid:x}
+async ID        0x{aid:x}
+tree ID         0x{tid:x}
+signature       {sig}""",
+            pid=protocol_id,
+            hs=packet.hdr.size,
+            cc=packet.hdr.credit_charge,
+            status=packet.hdr.status,
+            cmd=COMMANDS[packet.hdr.command][0],
+            cmdn=packet.hdr.command,
+            cr=packet.hdr.credit_request,
+            flags=packet.hdr.flags,
+            flags_desc=flags_desc,
+            nc=packet.hdr.next_command,
+            mid=packet.hdr.message_id,
+            sid=packet.hdr.session_id,
+            aid=packet.hdr.async_id,
+            tid=packet.hdr.tree_id,
+            sig=binascii.hexlify(packet.hdr.signature),
+        )
+        if packet.hdr.command < len(COMMANDS):
+            name, req_type, resp_type = COMMANDS[packet.hdr.command]
+            func = "smb_" + name
+            try:
+                if func in globals() and req_type:
```

Comment:
Dispatch out of `globals()` is potentially quite slow (`globals()`, `locals()`, `currentframe()` and the like are debugging and metaprogramming features, not the sort of thing you should be calling in a tight loop like this) and error prone; it's hard to debug when something goes wrong with this.  Not to mention that the type checker has no visibility here.

Luckily, here you already control all the relevant namespaces, so you don't need this level of dynamic metaprogramming! Define `COMMANDS` at the bottom of the file, and just add the functions to it directly, then do `name, reqType, respType, handler = COMMANDS[packet.hdr.command]`.